trillian/ctdns/handlers.go

```diff
@@ -0,0 +1,311 @@
+	defer cancelFunc()
+	req := &trillian.GetLeavesByHashRequest{
+		LogId:    c.cfg.LogId,
+		LeafHash: [][]byte{h},
+	}
+	resp, err := c.client.GetLeavesByHash(ctx, req)
+	if err != nil || len(resp.Leaves) != 1 {
+		glog.Warningf("hashFunc(): GetLeavesByHashRequest=%v, %v", resp, err)
+		failWithRcode(w, r, dns.RcodeServerFailure)
+		return
+	}
+	rr := buildHashResponse(params[0], resp.Leaves[0])
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Answer = append(m.Answer, rr)
+	if err := w.WriteMsg(m); err != nil {
+		glog.Warningf("hashFunc(): WriteMsg: %v", err)
+	}
+}
+
+// params will contain 0 = regex match text, 1 = start index, 2 = leaf_index,
+// 3 = tree_size, 4 = zone.
+func treeFunc(c *CTDNSHandler, params []string, w dns.ResponseWriter, r *dns.Msg) {
+	// We don't expect these to fail as the regex matched digits but check anyway.
+	values, err := parseInts(params, 1, 4)
+	if err != nil {
+		failWithRcode(w, r, dns.RcodeServerFailure)
+		return
+	}
+	ctx, cancelFunc := context.WithTimeout(context.Background(), c.opts.Deadline)
+	defer cancelFunc()
+	req := &trillian.GetInclusionProofRequest{
+		LogId:     c.cfg.LogId,
+		LeafIndex: values[2],
+		TreeSize:  values[3],
+	}
+	resp, err := c.client.GetInclusionProof(ctx, req)
+	if err != nil {
+		glog.Warningf("hashFunc(): GetInclusionProof=%v, %v", resp, err)
+		failWithRcode(w, r, dns.RcodeServerFailure)
+		return
+	}
+	// Ensure the client requested a valid start index for the proof.
+	if values[1] < 0 || values[1] >= int64(len(resp.Proof.Hashes)) {
+		failWithRcode(w, r, dns.RcodeServerFailure)
+		return
+	}
+	rr := buildProofResponse(params[0], int(values[1]), resp.GetProof())
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Answer = append(m.Answer, rr)
+	if err := w.WriteMsg(m); err != nil {
+		glog.Warningf("treeFunc(): WriteMsg: %v", err)
+	}
+}
+
+func buildHashResponse(q string, l *trillian.LogLeaf) dns.RR {
+	// Response has one element, the leaf index.
+	rr := &dns.TXT{
+		Hdr: dns.RR_Header{Name: dns.Fqdn(q), Class: dns.ClassINET, Rrtype: dns.TypeTXT, Ttl: 0},
+		Txt: []string{fmt.Sprintf("%d", l.LeafIndex)},
+	}
+
+	return rr
+}
+
+func buildProofResponse(q string, s int, proof *trillian.Proof) dns.RR {
+	var p []byte
+	// We can pack a limited number of proof elements into the dns response.
+	for i := s; i < len(proof.Hashes); i++ {
+		if len(p)+len(proof.Hashes[i]) > maxProofLen {
+			break
+		}
+		p = append(p, proof.Hashes[i]...)
+	}
+	// Response has one element. Note that the response is binary and not encoded.
+	rr := &dns.TXT{
+		Hdr: dns.RR_Header{Name: dns.Fqdn(q), Class: dns.ClassINET, Rrtype: dns.TypeTXT, Ttl: 0},
+		Txt: []string{string(p)},
+	}
+
+	return rr
+}
+
+func buildSTHResponse(q string, root *types.LogRootV1) dns.RR {
+	rh := base64.StdEncoding.EncodeToString(root.RootHash)
+	ts := root.TimestampNanos / 1000 / 1000 // Convert to millis.
+	ths := "todo"                           // We don't store this so will need to re-sign like CTFE
```

Comment:
Could we make this a proper TODO, and also return something that's valid base64 (so clients will reject for invalid-signature rather than for malformed response).

Reply:
I'm in the middle of implementing it. This wasn't really ready for review yet.

Reply:
Done.